apps/heft/src/configuration/HeftParameterManager.ts

```diff
@@ -0,0 +1,166 @@
+import { InternalError } from '@rushstack/node-core-library';
+import type { CommandLineParameter, CommandLineParameterProvider } from '@rushstack/ts-command-line';
+
+import type {} from '../pluginFramework/IHeftPlugin';
+import type {
+  HeftPluginDefinitionBase,
+  IChoiceParameterAlternativeJson,
+  IParameterJson
+} from './HeftPluginDefinition';
+
+export interface IParameterConfiguration {
+  pluginDefinition: HeftPluginDefinitionBase;
+  parameter: IParameterJson;
+}
+
+export class HeftParameterManager {
+  private _isFinalized: boolean = false;
+  private _parameterConfigurations: Set<IParameterConfiguration> = new Set();
+  // plugin definition => Map< parameter long name => applied parameter >
+  private _parametersByDefinition: Map<HeftPluginDefinitionBase, Map<string, CommandLineParameter>> =
+    new Map();
+
+  public addPluginParameters(pluginDefinition: HeftPluginDefinitionBase): void {
+    if (this._isFinalized) {
+      throw new InternalError('Parameters have already been finalized.');
+    }
+
+    // Parameters from this plugin have already been added
+    if (this._parametersByDefinition.has(pluginDefinition)) {
+      return;
+    } else {
+      this._parametersByDefinition.set(pluginDefinition, new Map());
+    }
+
+    // Add the parameters provided by the plugin. We will perform parameter conflict resolution when applying
```

Comment:
I don't see any parameter conflict resolution, unless you mean it occurs when defining the parameters internally.